### PR TITLE
[WIP]cmd, core: resolve scheme from a read-write database

### PR DESCRIPTION
**Pull request by sysvm**
### Description

This pull request solves four pull requests: [#2110](https://github.com/bnb-chain/bsc/issues/2110), [#2131](https://github.com/bnb-chain/bsc/issues/2131),  [#2141](https://github.com/bnb-chain/bsc/issues/2141), [#2146](https://github.com/bnb-chain/bsc/issues/2146)

1. The first question: resolve state scheme from a read-write database:

This issue can occur when unclean shutdown happens and freezer tables need to be repair. In database read-only mode, the freezer will only be validated but not repair and terminates the node abnormally.
This pull request moves the scheme check into the Ethereum constructor where the read-write database is already initialized. This approach can avoid opening read-only database in very early stage, to prevent this error totally.

2. The second question: ancient directory should be compatible with hash and path mode:

Current geth doesn't deal with chain freezer dir in a good way. Therefore, this pr refactors `resolveChainFreezerDir` to fix in pruned and non-pruned mode.

### Rationale

These will influence how should users use `pruenancient` mode that block data could be replaced in which.

### Example

N/A

### Changes

Notable changes: 
* N/A

### CODE REVIEW

**Review by Fynnss — DISMISSED** [review]
LGTM

File: cmd/geth/chaincmd.go [hunk 1]
```diff
@@ -655,7 +655,7 @@ func dump(ctx *cli.Context) error {
 	if err != nil {
 		return err
 	}
-	triedb := utils.MakeTrieDatabase(ctx, db, true, false) // always enable preimage lookup
```
Comment (Fynnss) [inline cmd/geth/chaincmd.go:658]: Why not readonly mode? It's dangerous.

Comment (sysvm) [reply cmd/geth/chaincmd.go:658]: Original code is not read-only mode. I change it into read-only mode.

File: core/rawdb/database.go [hunk 2]
```diff
@@ -219,12 +219,26 @@ func NewFreezerDb(db ethdb.KeyValueStore, frz, namespace string, readonly bool,
...
 	// - chain freezer is not initialized
 	// - chain freezer exists in legacy location (root ancient folder)
 	freezer := path.Join(ancient, chainFreezerName)
+	state := path.Join(ancient, stateFreezerName)
+	if pruned {
+		if common.FileExist(freezer) {
+			return freezer
+		}
+		if common.FileExist(state) {
+			return freezer
```
Comment (zzzckck) [inline core/rawdb/database.go:234]: not state?

**Comment by buddh0** [comment]
some PR are from go-ethereum
so I think, we can format this PR:
PR from go-ethereum use cherry-pick and solve conflict
then apply our commit

**Comment by zzzckck** [comment]
LGTM, would approve once updated to keep the origin commit record.

**Pull request closed without merging**